File: arima.py
from pandas import read_csv
from pandas import datetime
from pandas import DataFrame

import math
import pandas as pd

from processing import *

from pandas.tools.plotting import autocorrelation_plot
from statsmodels.tsa.arima_model import ARIMA

#dataframe = read_csv('ibov_google_15jun2017_1min_15d.csv', sep = ',', usecols=[1],
#  engine='python', skiprows=8, decimal='.',header=None)

dataframe = read_csv('minidolar/wdo.csv', sep = '|', usecols=[5],  engine='python', decimal='.',header=0)
dataframe = dataframe['fechamento']
dataframe = dataframe[0:540]


# BTC-USD
# btc = read_csv('btc-usd.csv', sep = ',',  engine='python', decimal='.',header=0)
# dataframe = btc['close']


start_time = time.time()


# test_stationarity(dataframe)
#
# plt.plot(dataframe.values)
# autocorrelation_plot(dataframe)
# plt.show()


# model = ARIMA(dataframe.as_matrix(), order=(5,1,0))
# model_fit = model.fit(disp=0)
# print(model_fit.summary())
# # plot residual errors
# residuals = DataFrame(model_fit.resid)
# residuals.plot()
# plt.show()
# residuals.plot(kind='kde')
# plt.show()
# print(residuals.describe())



# DOLLAR ARIMA(0,1,1)
# BTC ARIMA(3,1,5)

p=2
d=1
q=1

STEP = 1
FORECAST = 1

predictions,test = [],[]
for i in range(int(len(dataframe)-11), len(dataframe), STEP):
	try:
		x_i = np.asarray(dataframe[0:i]) #ARIMA have to use all timeseries
		y_i = dataframe[i+FORECAST]
		model = ARIMA(x_i, order=(p,d,q))
		model_fit = model.fit(disp=0)
		output = model_fit.forecast()
		yhat = output[0][0]
	except Exception as e:
		break

	predictions.append(yhat)
	test.append(y_i)

error = math.sqrt(mean_squared_error(test, predictions))
print('Test RMSE: %.3f' % error)
elapsed_time = (time.time() - start_time)
with open("output/arima.csv", "a") as fp:
	fp.write("p %i, d %i, q %i, rmse %f --%s seconds\n" % (p, d, q, error, elapsed_time))
# plot
#plt.plot(test)
#plt.plot(predictions, color='red')
#plt.show()

File: arima.R
# install.packages("TSPred")
library(TSPred)

#
# # MINI Dollar
#
# start_time <- Sys.time()
#
# dolar = read.csv('minidolar/wdo.csv', sep = '|')
# dolar_train = dolar$fechamento[1:round(0.8*nrow(dolar))]
# dolar_test = dolar$fechamento[round(0.8*nrow(dolar)+1):nrow(dolar)]
#
# xx = dolar$fechamento
# currentIndex<-length(dolar_train);
# len = NROW( dolar );
# forecasts_arima_fit = dolar_test;
# repeat
# {
#   nextIndex = currentIndex + 1
#
#   # Get the series
#   yy = xx[1:currentIndex]
#   yp = c(xx[nextIndex])
#
#   fittest <- fittestArima(yy, yp)
#
#   # Save the forecast
#   forecasts_arima_fit[currentIndex-length(dolar_train)+1] = fittest$pred$pred[1]
#
#
#   if( nextIndex > len -1) break
#
#   currentIndex = nextIndex
# }
#
#
# MSE.arima <- MSE(dolar_test,forecasts_arima_fit)
# RMSE.arima <- sqrt(MSE.arima)
# print('MINI Dollar')
# print(RMSE.arima)
#
# end_time <- Sys.time()
#
# elapsed_time = end_time - start_time
#
# writeLines(paste('MINI Dollar',RMSE.arima,elapsed_time, sep=','), 'output/arimaR.csv')
#
# # BTC-USD
#
#
# btc = read.csv('btc-usd.csv', sep = ',')
# btc_train = btc$close[1:round(0.8*nrow(btc))]
# btc_test = btc$close[round(0.8*nrow(btc)+1):nrow(btc)]
#
# xx = btc$close
# currentIndex<-length(btc_train);
# len = NROW( btc );
# forecasts_arima_fit = btc_test;
#
# repeat
# {
#   nextIndex = currentIndex + 1
#
#   # Get the series
#   yy = xx[1:currentIndex]
#   yp = c(xx[nextIndex])
#
#   fittest <- fittestArima(yy, yp)
#
#   # Save the forecast
#   forecasts_arima_fit[currentIndex-length(btc_train)+1] = fittest$pred$pred[1]
#
#
#   if( nextIndex > len -1) break
#
#   currentIndex = nextIndex
# }
#
#
# MSE.arima <- MSE(btc_test,forecasts_arima_fit)
# RMSE.arima <- sqrt(MSE.arima)
# print('BTC-USD')
# print(RMSE.arima)
#
# end_time <- Sys.time()
#
# elapsed_time = end_time - start_time
#
# writeLines(paste('BTC-USD',RMSE.arima,elapsed_time, sep=','), 'output/arimaR.csv')


# furnas


btc = read.csv('furnas-vazoes-medias-mensais-m3s.csv', sep = ',')
btc_train = btc$furnas[1:round(0.8*nrow(btc))]
btc_test = btc$furnas[round(0.8*nrow(btc)+1):nrow(btc)]

xx = btc$furnas
currentIndex<-length(btc_train);
len = NROW( btc );
forecasts_arima_fit = btc_test;

repeat
{
  nextIndex = currentIndex + 1

  # Get the series
  yy = xx[1:currentIndex]
  yp = c(xx[nextIndex])

  fittest <- fittestArima(yy, yp)

  # Save the forecast
  forecasts_arima_fit[currentIndex-length(btc_train)+1] = fittest$pred$pred[1]


  if( nextIndex > len -1) break

  currentIndex = nextIndex
}


MSE.arima <- MSE(btc_test,forecasts_arima_fit)
RMSE.arima <- sqrt(MSE.arima)
print('furnas')
print(RMSE.arima)

end_time <- Sys.time()

elapsed_time = end_time - start_time

writeLines(paste('furnas',RMSE.arima,elapsed_time, sep=','), 'output/arimaR.csv')



# rain


btc = read.csv('annual-rainfall-at-fortaleza-bra.csv', sep = ',')
btc_train = btc$rainfall[1:round(0.8*nrow(btc))]
btc_test = btc$rainfall[round(0.8*nrow(btc)+1):nrow(btc)]

xx = btc$rainfall
currentIndex<-length(btc_train);
len = NROW( btc );
forecasts_arima_fit = btc_test;

repeat
{
  nextIndex = currentIndex + 1

  # Get the series
  yy = xx[1:currentIndex]
  yp = c(xx[nextIndex])

  fittest <- fittestArima(yy, yp)

  # Save the forecast
  forecasts_arima_fit[currentIndex-length(btc_train)+1] = fittest$pred$pred[1]


  if( nextIndex > len -1) break

  currentIndex = nextIndex
}


MSE.arima <- MSE(btc_test,forecasts_arima_fit)
RMSE.arima <- sqrt(MSE.arima)
print('furnas')
print(RMSE.arima)

end_time <- Sys.time()

elapsed_time = end_time - start_time

writeLines(paste('furnas',RMSE.arima,elapsed_time, sep=','), 'output/arimaR.csv')